benchmarks/src/main/kotlin/benchmarks/bfschannel/BfsChannelBenchmark.kt

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+@file:JvmName("BfsChannelBenchmark")
+
+package benchmarks.bfschannel
+
+
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.nield.kotlinstatistics.standardDeviation
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.InputStreamReader
+import java.io.PrintWriter
+import java.net.URL
+import java.nio.channels.Channels
+import java.nio.file.Paths
+import java.util.*
+import java.util.concurrent.atomic.AtomicLong
+import java.util.zip.GZIPInputStream
+import kotlin.collections.ArrayList
+
+private val GRAPHS = listOf(
+        RandomGraphCreator("RAND-1M-10M", 1_000_000, 10_000_000),
+        GrGzGraphCreator("USA-DISTANCE", "http://www.dis.uniroma1.it/challenge9/data/USA-road-d/USA-road-d.USA.gr.gz"),
+        TxtGzGraphCreator("LIVE-JOURNAL", "https://snap.stanford.edu/data/soc-LiveJournal1.txt.gz"))
+/**
+ * Iterations count for each graph
+ */
+private const val ITERATIONS = 5
+/**
+ * Number of coroutines that are used to execute bfs in parallel
+ */
+private val COROUTINES = listOf(1, 4, 8, 16)
+/**
+ * Output file for the benchmark results
+ */
+private const val RESULT_FILE = "out/resultsBfs.csv"
+
+/**
+ * This benchmark tests channel as a working queue, as a queue under contention.
+ * The benchmark creates or downloads graphs, then executes parallel BFS using channel as a queue, executes sequential BFS,
+ * compares the results and computes execution times.
+ *
+ * NB: this benchmark works painfully slow without syncronization on methods [kotlinx.coroutines.internal.LockFreeLinkedListNode.remove]
+ * and [kotlinx.coroutines.internal.LockFreeLinkedListNode.helpDelete] (or other channels fixes).
+ */
+fun main() {
+    val results = ArrayList<String>()
+    for (graphCreator in GRAPHS) {
+        val graphName = graphCreator.name
+        println("=== $graphName ===")
+        val graph = graphCreator.getGraph()
+
+        val startNode = graph[0]
+
+        // warmup
```

Comment:
You have the same logic for both sequential and parallel algorithms, both of them should return `ExecutionResult`. Let's write an inline function which abstracts this running code/